File: apps/api/backend/services/editions.py
"""Family / edition parsing from documents.name.

family_key is only a scope limiter: a wrong key can under-merge, never make
propagation unsound.
"""

from __future__ import annotations

import re
from typing import Any, Dict, Optional, Tuple

_DATE_PATTERNS = [
    re.compile(
        r"(?:amended|upto|up\s*to|as\s+on|dated).{0,40}?"
        r"(\d{1,2}[./-]\d{1,2}[./-]\d{2,4})",
        re.I,
    ),
    re.compile(
        r"(?:amended|upto|up\s*to|as\s+on|dated).{0,40}?\b((?:19|20)\d{2})\b",
        re.I,
    ),
    re.compile(r"\b(20\d{2}|19\d{2})\s*[-–]\s*(\d{2})\b"),
    re.compile(r"\b(20\d{2}|19\d{2})\b"),
]

# Canonical display titles for well-known statute families (after normalization).
_CANONICAL_FAMILIES = (
    (re.compile(r"^income\s+tax\s+ordinance(?:\s*,?\s*2001)?$", re.I), "income tax ordinance, 2001"),
    (re.compile(r"^customs\s+act(?:\s*,?\s*1969)?$", re.I), "customs act, 1969"),
    (re.compile(r"^sales\s+tax\s+act(?:\s*,?\s*1990)?$", re.I), "sales tax act, 1990"),
    (re.compile(r"^federal\s+excise\s+act(?:\s*,?\s*2005)?$", re.I), "federal excise act, 2005"),
)


def family_key_from_name(name: str) -> str:
    raw = (name or "").strip()
    if not raw:
        return "unknown"
    base = re.sub(r"\(.*?\)", " ", raw)
    # ITO style: "… 2001 - amended upto …" — drop dash before amended/upto.
    base = re.sub(r"\s*[-–]\s*(?=(?:as\s+)?amended|upto|up\s*to|dated)", " ", base, flags=re.I)
    base = re.sub(r",?\s*(as\s+)?amended.*$", " ", base, flags=re.I)
    base = re.sub(r",?\s*upto.*$", " ", base, flags=re.I)
    base = re.sub(r",?\s*up\s*to.*$", " ", base, flags=re.I)
    base = re.sub(r",?\s*dated.*$", " ", base, flags=re.I)
    base = re.sub(r"^the\s+", " ", base, flags=re.I)
    base = re.sub(r"\s*,\s*", ", ", base)
    base = re.sub(r"\s+", " ", base).strip().rstrip(",").strip()
    base = base.lower() or "unknown"

    for pattern, canonical in _CANONICAL_FAMILIES:
        if pattern.match(base):
            return canonical
    # Finance Acts: collapse "finance act, 2025" → keep year out of family for grouping
    # by stripping trailing year when the title is Finance Act / Finance Supplementary Act.
    m = re.match(
        r"^(finance(?:\s+supplementary)?\s+act),?\s*(?:19|20)\d{2}(?:\s*[-–]\s*\d{2})?$",
        base,
        re.I,
    )
    if m:
        return m.group(1).lower()
    return base


def family_title_from_key(family_key: str) -> str:
    """Human title from a normalized family key."""
    raw = (family_key or "").strip()
    if not raw or raw == "unknown":
        return "Unknown statute"
    return raw[:1].upper() + raw[1:] if raw else "Unknown statute"


def edition_date_from_name(name: str) -> Dict[str, Any]:
    raw = name or ""
    for cre in _DATE_PATTERNS:
        m = cre.search(raw)
        if not m:
            continue
        if m.lastindex and m.lastindex >= 2 and m.group(1) and len(m.group(1)) == 4:
            year = int(m.group(1))
            return {"year": year, "sort_key": year, "label": str(year), "unknown": False}
        token = m.group(1)
        if re.match(r"^\d{1,2}[./-]\d{1,2}[./-]\d{2,4}$", token):
            parts = re.split(r"[./-]", token)
            year = int(parts[2])
            if year < 100:
                year += 2000
            return {"year": year, "sort_key": year, "label": str(year), "unknown": False}
        year = int(token)
        if 1900 <= year <= 2100:
            return {"year": year, "sort_key": year, "label": str(year), "unknown": False}
    return {"year": None, "sort_key": 9999, "label": "year unknown", "unknown": True}


def family_and_year(name: str) -> Tuple[str, Optional[int], int]:
    fk = family_key_from_name(name)
    ed = edition_date_from_name(name)
    return fk, ed["year"], int(ed["sort_key"])

File: apps/web/src/utils/editions.js
/**
 * Edition / family helpers derived from documents.name.
 * Family key is a scope limiter only — wrong keys under-merge, never over-merge.
 */

const DATE_PATTERNS = [
    /(?:amended|upto|up\s*to|as\s+on|dated).{0,40}?(\d{1,2}[./-]\d{1,2}[./-]\d{2,4})/i,
    /(?:amended|upto|up\s*to|as\s+on|dated).{0,40}?\b((?:19|20)\d{2})\b/i,
    /\b(20\d{2})\s*[-–]\s*(\d{2})\b/, // 2011-12 — before bare year
    /\b(20\d{2}|19\d{2})\b/,
];

const CANONICAL_FAMILIES = [
    [/^income\s+tax\s+ordinance(?:\s*,?\s*2001)?$/i, 'income tax ordinance, 2001'],
    [/^customs\s+act(?:\s*,?\s*1969)?$/i, 'customs act, 1969'],
    [/^sales\s+tax\s+act(?:\s*,?\s*1990)?$/i, 'sales tax act, 1990'],
    [/^federal\s+excise\s+act(?:\s*,?\s*2005)?$/i, 'federal excise act, 2005'],
];

export function familyKeyFromName(name) {
    const raw = String(name || '').trim();
    if (!raw) return 'unknown';
    let base = raw
        .replace(/\(.*?\)/g, ' ')
        .replace(/\s*[-–]\s*(?=(?:as\s+)?amended|upto|up\s*to|dated)/i, ' ')
        .replace(/,?\s*(as\s+)?amended.*$/i, ' ')
        .replace(/,?\s*upto.*$/i, ' ')
        .replace(/,?\s*up\s*to.*$/i, ' ')
        .replace(/,?\s*dated.*$/i, ' ')
        .replace(/^the\s+/i, ' ')
        .replace(/\s*,\s*/g, ', ')
        .replace(/\s+/g, ' ')
        .trim()
        .replace(/[,\s]+$/g, '')
        .trim()
        .toLowerCase();
    if (!base) return 'unknown';

    for (const [pattern, canonical] of CANONICAL_FAMILIES) {
        if (pattern.test(base)) return canonical;
    }

    const finance = base.match(
        /^(finance(?:\s+supplementary)?\s+act),?\s*(?:19|20)\d{2}(?:\s*[-–]\s*\d{2})?$/i,
    );
    if (finance) return finance[1].toLowerCase();

    return base;
}

export function familyTitleFromKey(familyKey) {
    const raw = String(familyKey || '').trim();
    if (!raw || raw === 'unknown') return 'Unknown statute';
    return raw.charAt(0).toUpperCase() + raw.slice(1);
}

/**
 * @returns {{ year: number | null, sortKey: number, label: string, unknown: boolean }}
 */
export function editionDateFromName(name) {
    const raw = String(name || '');
    for (const re of DATE_PATTERNS) {
        const m = raw.match(re);
        if (!m) continue;
        if (m[2] && m[1] && m[1].length === 4) {
            // 2011-12 style — use first year
            const year = Number(m[1]);
            return { year, sortKey: year, label: String(year), unknown: false };
        }
        const token = m[1];
        if (/^\d{1,2}[./-]\d{1,2}[./-]\d{2,4}$/.test(token)) {
            const parts = token.split(/[./-]/);
            let year = Number(parts[2]);
            if (year < 100) year += 2000;
            return { year, sortKey: year, label: String(year), unknown: false };
        }
        const year = Number(token);
        if (year >= 1900 && year <= 2100) {
            return { year, sortKey: year, label: String(year), unknown: false };
        }
    }
    return { year: null, sortKey: 9999, label: 'year unknown', unknown: true };
}

/**
 * The edition of a whole document, preferring the year the SERVER sorted by.
 *
 * `editionDateFromName` reads the name and prefers the year next to
 * "amended upto"; `YEAR_SQL` prefers `edition_date`, else the FIRST 19xx/20xx in
 * the name. For "Customs Act, 1969 as amended up to 30.06.2025" the server sorted
 * on 1969 and the badge read 2025 -- so "Edition — newest" produced an order that
 * contradicted the labels on the cards. Reading the server's value is what makes
 * them agree; the name is the fallback for a payload that carries no year.
 */
export function editionOf(doc) {
    const year = doc?.edition_year;
    if (typeof year === 'number' && Number.isFinite(year)) {
        return { year, sortKey: year, label: String(year), unknown: false };
    }
    return editionDateFromName(doc?.name || doc?.document_name || '');
}

export function sortEditions(docs) {
    return [...docs].sort((a, b) => {
        const da = editionOf(a);
        const db = editionOf(b);
        if (da.sortKey !== db.sortKey) return da.sortKey - db.sortKey;
        return String(a.name || '').localeCompare(String(b.name || ''));
    });
}
